### DISCUSSION TITLE
Future of the shaders branch

### DISCUSSION THREAD

**Discussion by kilacoda-old** [3b1b/manim#1310] [General]
The shaders branch seems to have been a work in progress for quite some time now, while the master branch hasn't seen any significant development (last commit in November 2020). It seems likely that the shaders branch is to be the future of this repo, but is there a timeline on it being merged with/replacing the master branch?

One reason I'm interested to know this is because there are people still using the master version here which has tons of bugs (a lot of which have been fixed in ManimCommunity/manim) and thus come asking questions on the same. I don't see the point in keeping a somewhat stale/dead branch as a default, while keeping the active branch in the shadows.

cc @3b1b

**Comment by TonyCrane** [comment]
In my opinion, the master branch here is currently a relatively stable version with no frequent updates. It may be convenient for some people to use, so leave it here. Although there are some bugs, it does not affect the use much (I am still using this version and made a lot of images and videos successfully without problems).
If one want more functions and solve some bugs, they can use the community version maintained by ManimCommunity.
The shaders branch will be merged into the master branch when it matures in the future, but it is not the time yet. And now we don’t put shaders on the default branch, I think it will be more beneficial for Grant to develop.

**Reply by kilacoda-old** [reply]
If it is a stable version, doesn't it make more sense to keep it as a secondary/semi-archived branch which serves as for example the base for the PyPI releases while keeping the in-development branch the default?
Also, out of curiosity, do you have an estimated timeframe in which you are considering to merge it?

**Reply by TonyCrane** [reply]
Sorry for not replying for a long time.
I prefer to treat manim as a software rather than a python package, so many practices may not be in line with conventional management of python packages.
At the same time, I recommend to always use the latest version on GitHub instead of the version on release or pypi, so I prefer to leave the stable version on the master branch here temporarily. And all the content in the shaders branch is under development by Grant, and there are still loopholes, so I think it will be a better way to continue to polish it in the shaders branch.
Currently the master branch only exists as a transition.
For the shaders branch, it can be used normally, but the content is still not perfect. I don't know when it will be merged into the master branch, but I don't think it will take a very long time.

**Comment by 3b1b** [comment]
It's now been merged into master.

In general, my loose plan for the future of this repo is to build out more functionality that sits on top of OpenGL, with new Mobject types that correspond to different shaders for different rendering styles.  At the moment, there are essentially four types, VMobject for vector graphics, Surface for 3d surfaces, DotCloud (which is meant to handle large numbers of points in a performant way), and ImageMobject.

There's some rudimentary functionality for editing the underlying shader program at a higher level, e.g. with Mobject.set_color_by_code, but it would be interesting to explore more of that so that the power of shaders can be blended with the ease and high-level reasoning of python.
